napari/_qt/_tests/test_qt_viewer.py

```diff
@@ -744,16 +744,21 @@ def _update_data(
 
 @pytest.fixture
 def qt_viewer_with_controls(qtbot):
+    def _before_close(qt_viewer_):
+        qt_viewer_.hide()
+        qt_viewer_._instances.clear()
+
+    def _before_close_controls(controls):
+        controls.hide()
+
     qt_viewer = QtViewer(viewer=ViewerModel())
+    qtbot.addWidget(qt_viewer, before_close_func=_before_close)
+    qtbot.addWidget(
+        qt_viewer.controls, before_close_func=_before_close_controls
+    )
     qt_viewer.show()
     qt_viewer.controls.show()
-    yield qt_viewer
-    qt_viewer.controls.hide()
-    qt_viewer.controls.close()
-    qt_viewer.hide()
-    qt_viewer.close()
-    qt_viewer._instances.clear()
-    qtbot.wait(50)
+    return qt_viewer
```

Comment:
These changes are super interesting! Are they equivalent except that instead of waiting a specified period of time, qtbot actually makes sure that the controls are closed?

Reply:
Yes. As I modified `make_napari_viewer` to not raise exception on `'handled_widget'` as leaked widget, we do not need to wait until every data is cleaned. We could allow `qtbot` to handle it in a common way.